### PR TITLE
Fixed #25185 -- Added support for functools.partial serialization in migrations

### PR DESCRIPTION
**Pull request by pipermerriam**
Code for https://code.djangoproject.com/ticket/25185

### What was wrong?

The `MigrationWriter` class did not support serialization of `functools.partial` objects.  These can be quite useful for use as the value for the `default` keyword for model fields.

### How was it fixed.

Since the `functools.partial` object has the three properties `func`, `args`, and `keywords`, the `MigrationWriter` now serializes these individual values and then reconstructs the call to `functools.partial`.

In addition, a special case was required for handling the serialization of functions that are in the `__builtins__` module.

#### Cute animal picture

the-world s-top-10-best-images-of-things-balanced-on-rabbits-heads-4

### CODE REVIEW

File: django/db/migrations/writer.py [hunk 1]
```diff
@@ -472,6 +490,8 @@ def serialize(cls, value):
                     "For more information, see "
                     "https://docs.djangoproject.com/en/%s/topics/migrations/#serializing-values"
                     % (value.__name__, module_name, get_docs_version()))
+            if module_name == '__builtin__':
```
Comment (carljm) [inline django/db/migrations/writer.py]: I guess this is not strictly related to the primary purpose of the commit (as in, the functools case would work without this)?

If so, it would be better to make this a separate pull request for clarity, and it also needs its own test.

File: django/db/migrations/writer.py [hunk 2]
```diff
@@ -439,6 +440,23 @@ def serialize(cls, value):
             string, imports = OperationWriter(value, indentation=0).serialize()
             # Nested operation, trailing comma is handled in upper OperationWriter._write()
             return string.rstrip(','), imports
+        elif isinstance(value, functools.partial):
+            imports = set(['import functools'])
```
Comment (charettes) [inline django/db/migrations/writer.py]: Use a `set` literal here.

File: django/db/migrations/writer.py (hunk 1 above)
Comment (pipermerriam) [inline django/db/migrations/writer.py]: Yes, you are correct.  I will open a separate ticket and pull request for this.

Do you have any suggestions on what function I could use in the test case?  The first one that comes to mind is `django.utils.crypto.get_random_string`.

File: django/db/migrations/writer.py (hunk 2 above)
Comment (pipermerriam) [inline django/db/migrations/writer.py]: 180dced2b110d7df8d566e33f093244937ef7110

File: tests/migrations/test_writer.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 import re
 import tokenize
 import unittest
+import functools
```
Comment (charettes) [inline tests/migrations/test_writer.py]: You'll need to order your imports to appease isort here; `functools` comes before `math`.

Comment (pipermerriam) [inline tests/migrations/test_writer.py]: 699f75373c4d9723366b218aab7c0bf0f2800056

File: django/db/migrations/writer.py (hunk 1 above)
Comment (carljm) [inline django/db/migrations/writer.py]: I must be missing something - `get_random_string` isn't a built-in function, so how could it be used to test this case?

I assumed since this code was in here, you had a real-world use case for needing to serialize a built-in function in a migration. Could that use case be turned into a test?

If not, maybe this isn't needed?

**Comment by pipermerriam** [comment]
@carljm if you take a look at d10af5b5dde363a19aa77d8f6d88408714135225 it might make it clear what I was doing a poor job communicating in my previous comment about `__builtins__`.  I was referencing needing a function to use to test the `functools.partial` serialization.  I encountered the `__builtins__` bug while attempting to serialize the `__builtins__.round` function, but looking back on it I think I was actually encountering a bug in how `WriterTests.safe_repr` executes code.

I've switched to using `datetime.timedelta` as my function that I'm passing into `functools.partial` in my test case and things seem to be working fine.

**Comment by carljm** [comment]
Ah, I get it now. Totally missed that you were using a builtin in your test for `partial`.

**Comment by carljm** [comment]
buildbut, test this please

**Comment by carljm** [comment]
buildbot, test this please

File: docs/topics/migrations.txt [hunk 4]
```diff
@@ -652,6 +652,7 @@ Django can serialize the following:
 - ``datetime.date``, ``datetime.time``, and ``datetime.datetime`` instances
   (include those that are timezone-aware)
 - ``decimal.Decimal`` instances
+- ``functools.partial`` instances which have serializable ``func``, ``args``, ``keywords`` values.
```
Comment (timgraham) [inline docs/topics/migrations.txt]: ", and keywords"
Please wrap line at 79 characters
Please add this below this list:
```
.. versionchanged:: 1.9

    Serialization support for `functools.partial` was added.
```
as well as a sentence in the 1.9 release notes.

**Comment by timgraham** [comment]
You may squash commits and use the commit message from the PR title when updating, thanks.

**Comment by pipermerriam** [comment]
> You may squash commits and use the commit message from the PR title when updating, thanks.

Done.  Thanks for the feedback @timgraham

**Comment by timgraham** [comment]
merged in 537818af8783a4f9f239ed6277b1f810c1666839, thanks!
